File: src/zvm/opcodes.js
/*

Z-Machine opcodes
=================

Copyright (c) 2011 The ifvms.js team
BSD licenced
http://github.com/curiousdannii/ifvms.js

*/

/*
	
TODO:
	consider generalising the indirect instructions
	Abstract out the signed conversions such that they can be eliminated if possible
	don't access memory directly
	consider scrapping the autowriter idea and calling the operands' .write()s directly
	
*/

// If the .write() of an operand generates code to access the stack or the memory (if non-native ByteArray) then don't access it more than once
// Currently only for branchers
var rUnsafeOperand = native_bytearrays ? /^s/ : /^[sm]/,
safe_operand = function( opcode, operand )
{
	var temp = operand();
	if ( rUnsafeOperand.test( temp ) )
	{
		pretemp( opcode, temp );
		temp = temp_var( opcode );
	}
	return temp;
},
// A temp var unique to this opcode
temp_var = function( opcode )
{
	return 't' + opcode.pc;
},
// Add a temporary var to the pre list
pretemp = function( opcode, value )
{
	opcode.pre.push( 'var ' + opcode.temp() + '=' + value );
},

// Z-Machine brancher
ZBrancher = Brancher.subClass({
	// Calculate the offset
	calc_offset: function()
	{
		var brancher = this.operands.pop(),
		word = brancher > 0xFF;
		this.offset = word ? ( (brancher & 0x1FFF) | (brancher & 0x2000 ? ~0x1FFF : 0) ) : brancher & 0x3F;
		this.iftrue = brancher & ( word ? 0x8000 : 0x80 );
	}
}),

// Common opcodes
alwaysbranch = opcode_builder( ZBrancher, function() { return 1; } ),

opcodes = {
	
/* je */ 1: opcode_builder( ZBrancher, function( a, b ) { return arguments.length == 2 ? a() + '==' + b() : 'e.jeq(' + this.var_args( arguments ) + ')'; } ),
/* jl */ 2: opcode_builder( ZBrancher, function( a, b ) { return a.U2S() + '<' + b.U2S(); } ),
/* jg */ 3: opcode_builder( ZBrancher, function( a, b ) { return a.U2S() + '>' + b.U2S(); } ),
/* dec_chk */
/* inc_chk */
/* jin */ 6: opcode_builder( ZBrancher, function( a, b ) { return 'e.jin(' + a() + ',' + b() + ')'; } ),
/* test */ 7: opcode_builder( ZBrancher, function( bitmap, flag ) { var temp = safe_operand( this, flag ); return bitmap() + '&' + temp + '==' + temp; } ),
/* or */ 8: opcode_builder( Storer, function( a, b ) { return a() + '|' + b(); } ),
/* and */ 9: opcode_builder( Storer, function( a, b ) { return a() + '&' + b(); } ),
/* test_attr */ 10: opcode_builder( ZBrancher, function( object, attr ) { return 'e.test_attr(' + object() + ',' + attr() + ')'; } ),
/* set_attr */
/* clear_attr */
/* store */ 13: opcode_builder( Indirect, function( variable, value ) { return value(); } ), // !!!
/* insert_obj */
/* loadw */ 15: opcode_builder( Storer, function( array, index ) { return 'm.getUint16(' + array() + '+2*' + index() + ')'; } ),
/* loadb */ 16: opcode_builder( Storer, function( array, index ) { return 'm.getUint8(' + array() + '+' + index() + ')'; } ),
/* get_prop */
/* get_prop_addr */ 18: opcode_builder( Storer, function( object, property ) { return 'e.get_prop_addr(' + object() + ',' + property() + ')'; } ),
/* get_next_prop */
/* add */ 20: opcode_builder( Storer, function( a, b ) { return 'e.S2U(' + a() + '+' + b() + ')'; } ),
/* sub */ 21: opcode_builder( Storer, function( a, b ) { return 'e.S2U(' + a() + '-(' + b() + '))'; } ),
/* mul */ 22: opcode_builder( Storer, function( a, b ) { return 'e.S2U(' + a() + '*' + b() + ')'; } ),
/* div */ 23: opcode_builder( Storer, function( a, b ) { return 'e.S2U(parseInt(' + a() + '/' + b() + '))'; } ),
/* mod */ 24: opcode_builder( Storer, function( a, b ) { return 'e.S2U(' + a() + '%' + b() + ')'; } ),
/* call_2s */ 25: CallerStorer,
/* call_2n */ 26: Caller,
/* set_colour */
/* throw */
/* jz */ 128: opcode_builder( ZBrancher, function( a ) { return a() + '==0'; } ),
/* get_sibling */
/* get_child */
/* get_parent */
/* get_prop_length */ 132: opcode_builder( Storer, function( a ) { return 'e.get_prop_len(' + a() + ')'; } ),
/* inc */ 133: opcode_builder( Opcode, function( a ) { return 'e.incdec(' + a() + ',1)'; } ),
/* dec */ 134: opcode_builder( Opcode, function( a ) { return 'e.incdec(' + a() + ',-1)'; } ),
/* print_addr */ 135: opcode_builder( Opcode, function( addr ) { return 'e.buffer+=e.text.decode(' + addr() + ')[0]' } ),
/* call_1s */ 136: CallerStorer,
/* remove_obj */
/* print_obj */ 138: opcode_builder( Opcode, function( a ) { return 'e.buffer+=e.text.decode(m.getUint16(e.objects+14*(' + a() + '-1)+13))[0]'; } ),
/* ret */ 139: opcode_builder( Stopper, function( a ) { return 'e.ret(' + a() + ')'; } ),
/* jump */ 140: opcode_builder( Stopper, function( a ) { return 'e.pc=' + a.U2S() + '+' + (this.next - 2) + ''; } ),
/* print_paddr */ 141: opcode_builder( Opcode, function( addr ) { return 'e.buffer+=e.text.decode(' + addr() + '*' + this.e.packing_multipler + ')[0]'; } ),
/* load */
/* call_1n */ 143: Caller,
/* rtrue */ 176: opcode_builder( Stopper, function() { return 'e.ret(1)'; } ),
/* rfalse */ 177: opcode_builder( Stopper, function() { return 'e.ret(0)'; } ),
// Reconsider a generalised class for @print/@print_ret?
/* print */ 178: opcode_builder( Opcode, function( text ) { return 'e.buffer+="' + text + '"'; }, { printer: 1 } ),
/* print_ret */ 179: opcode_builder( Stopper, function( text ) { return 'e.buffer+="' + text + '"'; }, { printer: 1 } ),
/* nop */ 180: Opcode,
/* restart */ 183: opcode_builder( Stopper, function() { return 'e.act("restart")'; } ), // !!!
/* ret_popped */ 184: Stopper.subClass({
	post: function() { this.operands.push( new Variable( this.e, 0 ) ); },
	func: function( a ) { return 'e.ret(' + a.write() + ')'; }
}),
/* catch */
/* quit */ 186: opcode_builder( Stopper, function() { return 'e.act("quit")'; } ),
/* new_line */ 187: opcode_builder( Opcode, function() { return 'e.buffer+="\\n"'; } ),
/* verify */ 189: alwaysbranch, // Actually check??
/* piracy */ 191: alwaysbranch,
/* call_vs */ 224: CallerStorer,
/* storew */ 225: opcode_builder( Opcode, function( array, index, value ) { return 'm.setUint16(' + array() + '+2*' + index() + ',' + value() + ')'; } ),
/* storeb */ 226: opcode_builder( Opcode, function( array, index, value ) { return 'm.setUint8(' + array() + '+' + index() + ',' + value() + ')'; } ),
/* put_prop */
/* aread */ 228: opcode_builder( Storer, function() { var storer = this.storer.v; this.storer = 0; return 'e.read(' + this.var_args( arguments ) + ',' + storer + ')'; }, { stopper: 1 } ),
/* print_char */ 229: opcode_builder( Opcode, function( a ) { return 'e.buffer+=String.fromCharCode(' + a() + ')'; } ),
/* print_num */ 230: opcode_builder( Opcode, function( a ) { return 'e.buffer+=' + a.U2S(); } ),
/* random */
/* push */ 232: Storer.subClass({
	storer: 0, // Don't grab an extra byte
	post: function() { this.storer = new Variable( this.e, 0 ); },
	func: function( a ) { return a.write(); }
}),
/* pull */
/* split_window */
/* set_window */
/* call_vs2 */ 236: CallerStorer,
/* erase_window */
/* erase_line */
/* set_cursor */
/* get_cursor */
/* set_text_style */ 241: opcode_builder( Opcode, function( stylebyte ) { return 'e.ui.set_style(' + stylebyte() + ')'; } ),
/* buffer_mode */
/* output_stream */
/* input_stream */
/* sound_effect */
/* read_char */
/* scan_table */
/* not */ 248: opcode_builder( Storer, function( a ) { return '~' + a(); } ),
/* call_vn */ 249: Caller,
/* call_vn2 */ 250: Caller,
/* tokenise */
/* encode_text */
/* copy_table */
/* print_table */
/* check_arg_count */
/* save */
/* restore */
/* log_shift */
/* art_shift */
/* set_font */
/* save_undo */
/* restore_undo */
/* print_unicode */ 1011: opcode_builder( Opcode, function( a ) { return 'e.buffer+=String.fromCharCode(' + a() + ')'; } ),
/* check_unicode */
// Assume we can print and read all unicode characters rather than actually testing
1012: opcode_builder( Storer, function() { return 3; } ),
/* gestalt */ 1030: opcode_builder( Storer, function() { return 'e.gestalt(' + this.var_args( arguments ) + ')'; } ),
/* parchment */ 1031: opcode_builder( Storer, function() { return 'e.op_parchment(' + this.var_args( arguments ) + ')'; } )
	
};

File: src/decompiler/ifvms-decompiler/src/zvm/opcodes.rs
/*

ZVM Opcodes
===========

Copyright (c) 2020 Dannii Willis
MIT licenced
https://github.com/curiousdannii/ifvms.js

*/

use fnv::FnvHashMap;

pub enum OperandTypes {
    Unsigned,
    Signed,
}
pub use OperandTypes::*;

pub struct OpcodeDefinition {
    pub stores: bool,
    pub branches: bool,
    pub ends_block: bool,
    pub pauses_vm: bool,
    pub operands: Option<Vec<OperandTypes>>,
}

fn simple_operands(operands: Vec<OperandTypes>) -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: false, ends_block: false, pauses_vm: false, operands: Some(operands)}
}

fn end_block() -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: false, ends_block: true, pauses_vm: false, operands: None}
}

fn pause_vm() -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: false, ends_block: true, pauses_vm: true, operands: None}
}

fn branch() -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: true, ends_block: false, pauses_vm: false, operands: None}
}

fn branch_and_pause_vm() -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: true, ends_block: true, pauses_vm: true, operands: None}
}

fn branch_with_operands(operands: Vec<OperandTypes>) -> OpcodeDefinition {
    OpcodeDefinition {stores: false, branches: true, ends_block: false, pauses_vm: false, operands: Some(operands)}
}

fn store() -> OpcodeDefinition {
    OpcodeDefinition {stores: true, branches: false, ends_block: false, pauses_vm: false, operands: None}
}

fn store_with_operands(operands: Vec<OperandTypes>) -> OpcodeDefinition {
    OpcodeDefinition {stores: true, branches: false, ends_block: false, pauses_vm: false, operands: Some(operands)}
}

fn store_and_branch() -> OpcodeDefinition {
    OpcodeDefinition {stores: true, branches: true, ends_block: false, pauses_vm: false, operands: None}
}

fn store_and_pause_vm() -> OpcodeDefinition {
    OpcodeDefinition {stores: true, branches: false, ends_block: true, pauses_vm: true, operands: None}
}

pub fn get_opcode_definitions(version: u8) -> FnvHashMap<u16, OpcodeDefinition> {
    let mut map = FnvHashMap::default();

    // Opcodes for all versions
    /* je */ map.insert(1, branch());
    /* jl */ map.insert(2, branch_with_operands(vec![Signed, Signed]));
    /* jg */ map.insert(3, branch_with_operands(vec![Signed, Signed]));
    /* dec_chk */ map.insert(4, branch_with_operands(vec![Unsigned, Signed]));
    /* inc_chk */ map.insert(5, branch_with_operands(vec![Unsigned, Signed]));
    /* jin */ map.insert(6, branch());
    /* test */ map.insert(7, branch());
    /* or */ map.insert(8, store());
    /* and */ map.insert(9, store());
    /* test_attr */ map.insert(10, branch());
    /* set_attr */
    /* clear_attr */
    /* store */
    /* insert_obj */
    /* loadw */ map.insert(15, store_with_operands(vec![Unsigned, Signed]));
    /* loadb */ map.insert(16, store_with_operands(vec![Unsigned, Signed]));
    /* get_prop */ map.insert(17, store());
    /* get_prop_addr */ map.insert(18, store());
    /* get_next_prop */ map.insert(19, store());
    /* add */ map.insert(20, store());
    /* sub */ map.insert(21, store());
    /* mul */ map.insert(22, store());
    /* div */ map.insert(23, store_with_operands(vec![Signed, Signed]));
    /* mod */ map.insert(24, store_with_operands(vec![Signed, Signed]));
    /* call_2s */ map.insert(25, end_block());
    /* call_2n */ map.insert(26, end_block());
    /* set_colour */
    /* throw */ map.insert(28, end_block());
    /* jz */ map.insert(128, branch());
    /* get_sibling */ map.insert(129, store_and_branch());
    /* get_child */ map.insert(130, store_and_branch());
    /* get_parent */ map.insert(131, store());
    /* get_prop_length */ map.insert(132, store());
    /* inc */
    /* dec */
    /* print_addr */
    /* call_1s */ map.insert(136, end_block());
    /* remove_obj */
    /* print_obj */
    /* ret */ map.insert(139, end_block());
    /* jump */ map.insert(140, OpcodeDefinition {stores: false, branches: false, ends_block: true, pauses_vm: false, operands: Some(vec![Signed])});
    /* print_paddr */
    /* load */ map.insert(142, store());
    /* rtrue */ map.insert(176, end_block());
    /* rfalse */ map.insert(177, end_block());
    /* print */
    /* print_ret */ map.insert(179, end_block());
    /* nop */
    /* restart */ map.insert(183, end_block());
    /* ret_popped */ map.insert(184, end_block());
    /* quit */ map.insert(186, pause_vm());
    /* new_line */
    /* verify */ map.insert(189, branch());
    /* piracy */ map.insert(191, branch());
    /* call_vs */ map.insert(224, end_block());
    /* storew */ map.insert(225, simple_operands(vec![Unsigned, Signed]));
    /* storeb */ map.insert(226, simple_operands(vec![Unsigned, Signed]));
    /* put_prop */
    /* print_char */
    /* print_num */ map.insert(230, simple_operands(vec![Signed]));
    /* random */ map.insert(231, store_with_operands(vec![Signed]));
    /* push */
    /* pull */
    /* split_window */
    /* set_window */
    /* call_vs2 */ map.insert(236, end_block());
    /* erase_window */ map.insert(237, simple_operands(vec![Signed]));
    /* erase_line */
    /* set_cursor */
    /* get_cursor */
    /* set_text_style */
    /* buffer_mode */
    /* output_stream */ map.insert(243, end_block());
    /* input_stream */ map.insert(244, pause_vm());
    /* sound_effect */
    /* read_char */ map.insert(246, store_and_pause_vm());
    /* scan_table */ map.insert(247, store_and_branch());
    /* not */ map.insert(248, store());
    /* call_vn */ map.insert(249, end_block());
    /* call_vn2 */ map.insert(250, end_block());
    /* tokenise */
    /* encode_text */
    /* copy_table */
    /* print_table */
    /* check_arg_count */ map.insert(255, branch());
    /* save */ map.insert(1000, store_and_pause_vm());
    /* restore */ map.insert(1001, store_and_pause_vm());
    /* log_shift */ map.insert(1002, store_with_operands(vec![Unsigned, Signed]));
    /* art_shift */ map.insert(1003, store_with_operands(vec![Signed, Signed]));
    /* set_font */ map.insert(1004, store());
    /* save_undo */ map.insert(1009, store());
    /* restore_undo */ // Need its store byte to be decoded, but don't set store
    /* print_unicode */
    /* check_unicode */ map.insert(1012, store());
    /* set_true_colour */
    /* gestalt */ map.insert(1030, store());

    if version < 4 {
        /* save */ map.insert(181, branch_and_pause_vm());
        /* restore */ map.insert(182, branch_and_pause_vm());
        /* show_status */ map.insert(188, pause_vm());
    } else {
        /* save */ map.insert(181, store_and_pause_vm());
        /* restore */ map.insert(182, store_and_pause_vm());
    }

    if version < 5 {
        /* not */ map.insert(143, store());
        /* pop */
        /* read */ map.insert(228, pause_vm());
    } else {
        /* call_1n */ map.insert(143, end_block());
        /* catch */ map.insert(185, store());
        /* read */ map.insert(228, store_and_pause_vm());
    }

    map
}